src/vm_cpu.rs

```diff
@@ -149,6 +160,120 @@ impl VmPmuState {
+        }
+    }
+
+    fn read_pmu_csr(csr: u64) -> Option<u64> {
+        #[inline]
+        fn read_csr<const V: u16>() -> u64 {
+            let r: u64;
+            // Safety: We have verified that these are valid PMU registers that exist on the platform
+            unsafe {
+                asm!("csrr {rd}, {csr}", rd = out(reg) r, csr = const V);
+            }
+            r
+        }
+
+        // These represent the range from CSR_CYCLE - CSR_HPMCOUNTER31
+        match csr {
+            0xC00 => Some(read_csr::<0xC00>()),
+            0xC01 => Some(read_csr::<0xC01>()),
+            0xC02 => Some(read_csr::<0xC02>()),
+            0xC03 => Some(read_csr::<0xC03>()),
+            0xC04 => Some(read_csr::<0xC04>()),
+            0xC05 => Some(read_csr::<0xC05>()),
+            0xC06 => Some(read_csr::<0xC06>()),
+            0xC07 => Some(read_csr::<0xC07>()),
+            0xC08 => Some(read_csr::<0xC08>()),
+            0xC09 => Some(read_csr::<0xC09>()),
+            0xC0A => Some(read_csr::<0xC0A>()),
+            0xC0B => Some(read_csr::<0xC0B>()),
+            0xC0C => Some(read_csr::<0xC0C>()),
+            0xC0D => Some(read_csr::<0xC0D>()),
+            0xC0E => Some(read_csr::<0xC0E>()),
+            0xC0F => Some(read_csr::<0xC0F>()),
+            0xC10 => Some(read_csr::<0xC10>()),
+            0xC11 => Some(read_csr::<0xC11>()),
+            0xC12 => Some(read_csr::<0xC12>()),
+            0xC13 => Some(read_csr::<0xC13>()),
+            0xC14 => Some(read_csr::<0xC14>()),
+            0xC15 => Some(read_csr::<0xC15>()),
+            0xC16 => Some(read_csr::<0xC16>()),
+            0xC17 => Some(read_csr::<0xC17>()),
+            0xC18 => Some(read_csr::<0xC18>()),
+            0xC19 => Some(read_csr::<0xC19>()),
+            0xC1A => Some(read_csr::<0xC1A>()),
+            0xC1B => Some(read_csr::<0xC1B>()),
+            0xC1C => Some(read_csr::<0xC1C>()),
+            0xC1D => Some(read_csr::<0xC1D>()),
+            0xC1E => Some(read_csr::<0xC1E>()),
+            0xC1F => Some(read_csr::<0xC1F>()),
+            _ => None,
+        }
+    }
+
+    /// Captures the current CSR value of started counters, after stopping them.
+    /// This is intended to be called in anticipation of out-bound context switch.
+    #[allow(dead_code)]
+    pub fn save_counters(&mut self) {
+        use drivers::pmu::PmuInfo;
+        use sbi::api::pmu;
+        let pmu_info = PmuInfo::get();
+        for counter_index in
+            self.started_counters.trailing_zeros() as u64..pmu_info.get_num_counters()
+        {
+            if self.started_counters & (1 << counter_index) != 0 {
+                let counter_index_mask = Self::gen_counter_mask(counter_index, counter_index);
+                // Ignore errors since we can't do anything about them.
+                let _ = pmu::stop_counters(
+                    counter_index,
+                    counter_index_mask,
+                    PmuCounterStopFlags::default(),
```

Comment:
Don't we need to un-configure the counters so that whatever VM we're switching to next can configure them?

Reply:
Don't believe that counters can be un-configured. Basically, the in-coming VM configures them, which is what we do in restore_counters(). @atishp04 Do you have any comments?

Reply:
Yup. see my previews comment

Reply:
> Don't believe that counters can be un-configured. Basically, the in-coming VM configures them, which is what we do in restore_counters().

Counters are unconfigured with `STOP_FLAG_RESET`. The incoming VM won't be able to reconfigure them if they are currently configured. That failure to reconfigure is currently hidden though due to the use of `CFG_FLAG_SKIP_MATCH`, per the below comment.